File: 2024/js/no1.js
/**
 * https://leetcode.cn/problems/two-sum/description/
 * no1
 * 两数之和
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = function(nums, target) {
  return hashTable(nums, target);
};
/**
 * 哈希表
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const hashTable = (nums, target) => {
  let hashT = [];
  let result = [];
  try {
    nums.forEach((x, index) => {
      let y = target - x;
      if (hashT[y] >= 0) {
        result.push(hashT[y]);
        result.push(index);
      }
      hashT[x] = index;
    });
    return result;
  } catch (e) {
    console.log(e);
    return result;
  }
}
/**
 * 遍历
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const traverse = (nums, target) => {
  let index = 0;
  for(let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if(nums[i] + nums[j] == target) {
        return [i,j];
      }
    }
  }
  return [];
};
console.log(twoSum([1, 2, 3], 4));

File: 2024/JavaSolution/src/main/java/com/bxom/javasolution/easy/NO1.java
package com.bxom.javasolution.easy;

import lombok.extern.slf4j.Slf4j;

import java.util.HashMap;
import java.util.Map;

@Slf4j
public class NO1 {

    /**
     * 哈希表
     *
     * @param nums
     * @param target
     * @return
     */
    private static int[] hashTable(int[] nums, int target) {
        Map<Integer, Integer> hTable = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int y = target - nums[i];
            if (hTable.containsKey(y)) {
                return new int[]{hTable.get(y), i};
            }
            hTable.put(nums[i], i);
        }
        return new int[]{};
    }

    /**
     * 嵌套遍历
     *
     * @param nums
     * @param target
     * @return
     */
    private static int[] traverse(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[]{i, j};
                }
            }
        }
        return new int[]{};
    }

    public static void main(String[] args) throws Exception {
        log.info("traverse result: {}", traverse(new int[]{1, 2, 3}, 4));
        log.info("hashTable result: {}", hashTable(new int[]{1, 2, 3}, 4));
    }
}
